modules/standard/CTypes.chpl

```diff
@@ -727,42 +725,145 @@ module CTypes {
+
+  /*
+    Returns a :type:`c_ptr` to the underlying buffer of a :type:`~String.string`
+
+    Note that the existence of this ``c_ptr`` has no impact on the lifetime of
+    the ``string``.  The returned pointer will be invalid if the ``string`` is
+    freed or even reallocated.
+
+    Halts if the ``string`` is empty and bounds checking is enabled.
+  */
+  inline proc c_ptrTo(ref s: string): c_ptrConst(c_uchar) {
+    if boundsChecking {
+      if (s.buffLen == 0) then
+        halt("Can't create a C pointer for an empty string.");
+    }
+    return c_pointer_return_const(s.buff[0]);
+  }
+
+  /*
+   Like ``c_ptrTo`` for :type:`~String.string`, but returns a :type:`c_ptrConst`
+   which disallows direct modification of the pointee.
+   */
+  inline proc c_ptrToConst(const ref s: string): c_ptrConst(c_uchar) {
+    if boundsChecking {
+      if (s.buffLen == 0) then
+        halt("Can't create a C pointer for an empty string.");
+    }
+    return c_pointer_return_const(s.buff[0]);
+  }
+
+  /*
+    Returns a :type:`c_ptr` to the underlying buffer of a :type:`~Bytes.bytes`
+
+    Note that the existence of this ``c_ptr`` has no impact on the lifetime of
+    the ``bytes``.  The returned pointer will be invalid if the ``bytes`` is
+    freed or even reallocated.
+
+    Halts if the ``bytes`` is empty and bounds checking is enabled.
+  */
+  inline proc c_ptrTo(ref b: bytes): c_ptrConst(c_uchar) {
+    if boundsChecking {
+      if (b.buffLen == 0) then
+        halt("Can't create a C pointer for an empty bytes.");
+    }
+    return c_pointer_return_const(b.buff[0]);
+  }
+
+  /*
+   Like ``c_ptrTo`` for :type:`~Bytes.bytes`, but returns a :type:`c_ptrConst`
+   which disallows direct modification of the pointee.
+   */
+  inline proc c_ptrToConst(const ref b: bytes): c_ptrConst(c_uchar) {
+    if boundsChecking {
+      if (b.buffLen == 0) then
+        halt("Can't create a C pointer for an empty bytes.");
+    }
+    return c_pointer_return_const(b.buff[0]);
+  }
+
   /*
     Returns a :type:`c_ptr` to any Chapel object.
     Note that the existence of the :type:`c_ptr` has no impact of the lifetime
     of the object. In many cases the object will be stack allocated and
     could go out of scope even if this :type:`c_ptr` remains.
 
-    :arg x: the by-reference argument to get a pointer to. The argument should
-            not be an array or domain (there is a different overload for arrays).
-            Records, class instances, integral, real, imag, and complex types are
-            supported.
+    :arg x: the by-reference argument to get a pointer to. Domains are not
+            supported, and will cause a compiler error. Records, class
+            instances, integral, real, imag, and complex types are supported.
+            For arrays, strings, or bytes, separate overloads should be used.
     :returns: a pointer to the argument passed by reference
 
   */
   inline proc c_ptrTo(ref x:?t):c_ptr(t) {
```

Comment:
I had imagined the body of the non-special-case `c_ptrTo[Const]` would just call the corresponding `c_addrOf[Const]` to avoid duplicating code. Is there a reason that doesn't work, or that this is preferable?

Reply:
That's a good idea, I'll change it.